File: packages/android/attachments-android/src/test/kotlin/app/notes/attachments/DefaultAttachmentPipelineTest.kt
package app.notes.attachments

import app.notes.domain.AttachmentId
import app.notes.domain.AttachmentRef
import app.notes.domain.FileInput
import app.notes.domain.NoteId
import app.notes.domain.Progress
import app.notes.domain.RemoteFile
import app.notes.domain.StorageProvider
import app.notes.domain.StorageProviderId
import app.notes.domain.UploadState
import app.notes.domain.UserId
import app.notes.domain.makeAttachmentRef
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class DefaultAttachmentPipelineTest {
    private val owner = UserId("550e8400-e29b-41d4-a716-446655440000")
    private val note = NoteId("550e8400-e29b-41d4-a716-446655440001")
    private val attId = AttachmentId("550e8400-e29b-41d4-a716-446655440010")
    private val clientId = "01234567-89ab-4cde-8f01-23456789abcd"

    private val pdfFile = FileInput(
        name = "spec.pdf",
        mimeType = "application/pdf",
        sizeBytes = 1024,
        bytes = ByteArray(1024),
    )
    private val pngFile = FileInput(
        name = "photo.png",
        mimeType = "image/png",
        sizeBytes = 2048,
        bytes = ByteArray(2048),
    )
    private val thumbnailFile = FileInput(
        name = "photo.thumb.jpg",
        mimeType = "image/jpeg",
        sizeBytes = 256,
        bytes = ByteArray(256),
    )

    private fun sampleRef(externalFileId: String = "drive-original", thumbnailId: String? = null): AttachmentRef {
        val builder = AttachmentRef(
            id = attId,
            ownerId = owner,
            noteId = note,
            provider = StorageProviderId.GoogleDrive,
            externalFileId = externalFileId,
            mimeType = "application/pdf",
            sizeBytes = 1024,
            displayName = "spec.pdf",
            thumbnailId = thumbnailId,
        )
        return builder
    }

    @Test
    fun `attach uploads bytes inserts row returns AttachmentRef`() = runTest {
        val storage = FakeStorage()
        val repo = FakeRepo(insertResult = sampleRef())
        val pipeline = DefaultAttachmentPipeline(storage, repo, newClientId = { clientId })
        val ref = pipeline.attach(note, pdfFile)
        assertEquals("drive-original", ref.externalFileId)
        assertEquals(1, storage.uploadCalls)
        assertEquals(1, repo.insertCalls)
    }

    @Test
    fun `attach generates thumbnail for image MIME types and uploads both`() = runTest {
        val storage = FakeStorage(uploadFn = { _: FileInput, hint: String? ->
            val id = if (hint == "thumbnail") "drive-thumb" else "drive-original"
            RemoteFile(id, "f", "image/png", 100)
        })
        val repo = FakeRepo(insertResult = sampleRef(thumbnailId = "drive-thumb"))
        val thumbs = ThumbnailGenerator { thumbnailFile }
        val pipeline = DefaultAttachmentPipeline(storage, repo, thumbs, newClientId = { clientId })
        val ref = pipeline.attach(note, pngFile)
        assertEquals("drive-thumb", ref.thumbnailId)
        assertEquals(2, storage.uploadCalls)
    }

    @Test
    fun `attach does NOT generate thumbnail for non-image MIME`() = runTest {
        val storage = FakeStorage()
        val repo = FakeRepo(insertResult = sampleRef())
        var thumbCalls = 0
        val thumbs = ThumbnailGenerator { thumbCalls += 1; thumbnailFile }
        val pipeline = DefaultAttachmentPipeline(storage, repo, thumbs, newClientId = { clientId })
        pipeline.attach(note, pdfFile)
        assertEquals(0, thumbCalls)
    }

    @Test
    fun `attach rolls back uploaded bytes when DB insert fails`() = runTest {
        val storage = FakeStorage()
        val repo = FakeRepo(insertResult = sampleRef(), insertThrows = RuntimeException("RLS denied"))
        val pipeline = DefaultAttachmentPipeline(storage, repo, newClientId = { clientId })
        assertThrows(RuntimeException::class.java) {
            kotlinx.coroutines.runBlocking { pipeline.attach(note, pdfFile) }
        }
        assertEquals(listOf("drive-original"), storage.deletedIds)
    }

    @Test
    fun `attach rolls back both files when DB insert fails after thumbnail upload`() = runTest {
        val storage = FakeStorage(uploadFn = { _: FileInput, hint: String? ->
            val id = if (hint == "thumbnail") "drive-thumb" else "drive-original"
            RemoteFile(id, "f", "image/png", 100)
        })
        val repo = FakeRepo(insertResult = sampleRef(), insertThrows = RuntimeException("RLS denied"))
        val thumbs = ThumbnailGenerator { thumbnailFile }
        val pipeline = DefaultAttachmentPipeline(storage, repo, thumbs, newClientId = { clientId })
        assertThrows(RuntimeException::class.java) {
            kotlinx.coroutines.runBlocking { pipeline.attach(note, pngFile) }
        }
        assertTrue("drive-original" in storage.deletedIds)
        assertTrue("drive-thumb" in storage.deletedIds)
    }

    @Test
    fun `attach proceeds with no thumbnail when thumbnail upload fails`() = runTest {
        var thumbAttempt = 0
        val storage = FakeStorage(uploadFn = { _: FileInput, hint: String? ->
            if (hint == "thumbnail") {
                thumbAttempt += 1
                throw RuntimeException("Drive 503 during thumbnail")
            }
            RemoteFile("drive-original", "f", "image/png", 100)
        })
        val repo = FakeRepo(insertResult = sampleRef())
        val thumbs = ThumbnailGenerator { thumbnailFile }
        val pipeline = DefaultAttachmentPipeline(storage, repo, thumbs, newClientId = { clientId })
        val ref = pipeline.attach(note, pngFile)
        assertEquals(1, thumbAttempt)
        assertNull(ref.thumbnailId)
        // The repo should have been called with thumbnailId = null.
        assertNull(repo.lastInsert?.thumbnailId)
    }

    @Test
    fun `attach proceeds when thumbnail GENERATOR fails`() = runTest {
        val storage = FakeStorage()
        val repo = FakeRepo(insertResult = sampleRef())
        val thumbs = ThumbnailGenerator { throw RuntimeException("Canvas exploded") }
        val pipeline = DefaultAttachmentPipeline(storage, repo, thumbs, newClientId = { clientId })
        val ref = pipeline.attach(note, pngFile)
        assertNull(ref.thumbnailId)
        assertEquals(1, storage.uploadCalls)
    }

    @Test
    fun `attach publishes Failed when original upload fails and repo insert is NOT called`() = runTest {
        val storage = FakeStorage(uploadFn = { _: FileInput, _: String? -> throw RuntimeException("Drive 503") })
        val repo = FakeRepo(insertResult = sampleRef())
        val pipeline = DefaultAttachmentPipeline(storage, repo, newClientId = { clientId })
        assertThrows(RuntimeException::class.java) {
            kotlinx.coroutines.runBlocking { pipeline.attach(note, pdfFile) }
        }
        assertEquals(0, repo.insertCalls)
    }

    @Test
    fun `uploadProgress ends in Done on successful attach`() = runTest {
        // Transition ordering (Pending → Uploading → Done) is verified on
        // the TS twin where the AsyncIterable contract is deterministic.
        // The Kotlin Flow + StandardTestDispatcher pair races on
        // subscribe-before-emit, so we verify the terminal state via the
        // bus's replay semantics instead.
        val storage = FakeStorage()
        val repo = FakeRepo(insertResult = sampleRef())
        val pipeline = DefaultAttachmentPipeline(storage, repo, newClientId = { clientId })
        pipeline.attach(note, pdfFile)
        val final = pipeline.uploadProgressByClientId(clientId).first()
        assertEquals(UploadState.Done, final.state)
    }

    @Test
    fun `uploadProgress ends in Failed on upload error`() = runTest {
        val storage = FakeStorage(uploadFn = { _: FileInput, _: String? -> throw RuntimeException("Drive 503") })
        val repo = FakeRepo(insertResult = sampleRef())
        val pipeline = DefaultAttachmentPipeline(storage, repo, newClientId = { clientId })
        assertThrows(RuntimeException::class.java) {
            kotlinx.coroutines.runBlocking { pipeline.attach(note, pdfFile) }
        }
        val final = pipeline.uploadProgressByClientId(clientId).first()
        assertTrue(final.state is UploadState.Failed)
    }

    @Test
    fun `resolveUrl proxies to storage with the externalFileId`() = runTest {
        val storage = FakeStorage()
        val pipeline = DefaultAttachmentPipeline(storage, FakeRepo(insertResult = sampleRef()))
        val url = pipeline.resolveUrl(sampleRef())
        assertEquals("https://drive/drive-original", url)
        assertEquals(listOf("drive-original"), storage.metadataCalls)
    }

    @Test
    fun `remove deletes DB row then remote bytes`() = runTest {
        val order = mutableListOf<String>()
        val storage = FakeStorage(deleteFn = { id -> order.add("storage:$id") })
        val repo = FakeRepo(insertResult = sampleRef(), deleteFn = { id -> order.add("db:${id.raw}") })
        val pipeline = DefaultAttachmentPipeline(storage, repo)
        pipeline.remove(sampleRef(thumbnailId = "drive-thumb"))
        assertEquals(
            listOf("db:${attId.raw}", "storage:drive-original", "storage:drive-thumb"),
            order,
        )
    }

    @Test
    fun `remove swallows storage delete failures`() = runTest {
        val storage = FakeStorage(deleteFn = { _ -> throw RuntimeException("Drive 503") })
        val repo = FakeRepo(insertResult = sampleRef())
        val pipeline = DefaultAttachmentPipeline(storage, repo)
        pipeline.remove(sampleRef())
        assertEquals(1, repo.deleteCalls)
    }

    @Test
    fun `progress bus replays the latest value to a late subscriber`() = runTest {
        val storage = FakeStorage()
        val repo = FakeRepo(insertResult = sampleRef())
        val pipeline = DefaultAttachmentPipeline(storage, repo, newClientId = { clientId })
        pipeline.attach(note, pdfFile)
        // Subscribe AFTER the upload completed; should get the final Done snapshot.
        val first = pipeline.uploadProgressByClientId(clientId).first()
        assertNotNull(first)
        assertEquals(UploadState.Done, first.state)
    }
}

private class FakeStorage(
    private val uploadFn: suspend (FileInput, String?) -> RemoteFile = { file, _ ->
        RemoteFile("drive-original", file.name, file.mimeType, file.sizeBytes)
    },
    private val deleteFn: suspend (String) -> Unit = { _ -> },
) : StorageProvider {
    override val id = StorageProviderId.GoogleDrive
    var uploadCalls = 0
    val deletedIds = mutableListOf<String>()
    val metadataCalls = mutableListOf<String>()

    override suspend fun upload(file: FileInput, folderHint: String?): RemoteFile {
        uploadCalls += 1
        return uploadFn(file, folderHint)
    }

    override suspend fun delete(externalId: String) {
        deletedIds.add(externalId)
        deleteFn(externalId)
    }

    override suspend fun getDownloadUrl(externalId: String, ttlSeconds: Int): String {
        metadataCalls.add(externalId)
        return "https://drive/$externalId"
    }

    override suspend fun ensureAppFolder(): String = "folder-1"
}

private class FakeRepo(
    private val insertResult: AttachmentRef,
    private val insertThrows: Throwable? = null,
    private val deleteFn: suspend (AttachmentId) -> Unit = { _ -> },
) : AttachmentsRepo {
    var insertCalls = 0
    var deleteCalls = 0
    var lastInsert: NewAttachmentRefInput? = null

    override suspend fun insert(input: NewAttachmentRefInput): AttachmentRef {
        insertCalls += 1
        lastInsert = input
        insertThrows?.let { throw it }
        // Re-shape the result to reflect the input thumbnailId (so tests
        // that pass through can observe it on the returned ref).
        return AttachmentRef(
            id = insertResult.id,
            ownerId = insertResult.ownerId,
            noteId = insertResult.noteId,
            provider = insertResult.provider,
            externalFileId = input.externalId,
            mimeType = input.mimeType,
            sizeBytes = input.sizeBytes,
            displayName = input.displayName,
            thumbnailId = input.thumbnailId,
        )
    }

    override suspend fun delete(id: AttachmentId) {
        deleteCalls += 1
        deleteFn(id)
    }

    override suspend fun listForValidation(cursor: String?, limit: Int): ValidationPage =
        ValidationPage(emptyList(), null)

    override suspend fun recordDangling(entry: DanglingEntry) {}
}

File: packages/attachments/src/__tests__/default-attachment-pipeline.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import {
  AttachmentId,
  NoteId,
  UserId,
  makeAttachmentRef,
  type AttachmentRef,
  type FileInput,
  type RemoteFile,
  type StorageProvider,
  type Progress,
} from "@notes-app/domain";
import { DefaultAttachmentPipeline } from "../default-attachment-pipeline.js";
import type { AttachmentsRepo, ThumbnailGenerator } from "../ports.js";

const OWNER = UserId("550e8400-e29b-41d4-a716-446655440000");
const NOTE = NoteId("550e8400-e29b-41d4-a716-446655440001");
const ATT = AttachmentId("550e8400-e29b-41d4-a716-446655440010");
const CLIENT_ID = "01234567-89ab-4cde-8f01-23456789abcd";

function sampleRef(overrides: Partial<AttachmentRef> = {}): AttachmentRef {
  return makeAttachmentRef({
    id: ATT,
    ownerId: OWNER,
    noteId: NOTE,
    provider: "GOOGLE_DRIVE",
    externalFileId: "drive-original",
    mimeType: "application/pdf",
    sizeBytes: 1024,
    displayName: "spec.pdf",
    ...overrides,
  });
}

function makeStorage(overrides: Partial<StorageProvider> = {}): StorageProvider {
  return {
    id: "GOOGLE_DRIVE",
    upload: vi.fn(async (file: FileInput) => makeRemote("drive-original", file)),
    delete: vi.fn(async () => {}),
    getDownloadUrl: vi.fn(async (id) => `https://drive/${id}`),
    ensureAppFolder: vi.fn(async () => "folder-1"),
    ...overrides,
  };
}

function makeRemote(externalId: string, file: FileInput): RemoteFile {
  return {
    externalId,
    name: file.name,
    mimeType: file.mimeType,
    sizeBytes: file.sizeBytes,
  };
}

function makeRepo(overrides: Partial<AttachmentsRepo> = {}): AttachmentsRepo {
  return {
    insert: vi.fn(async (input) =>
      sampleRef({
        externalFileId: input.externalId,
        sizeBytes: input.sizeBytes,
        mimeType: input.mimeType,
        displayName: input.displayName,
        ...(input.thumbnailId !== undefined ? { thumbnailId: input.thumbnailId } : {}),
      }),
    ),
    delete: vi.fn(async () => {}),
    listForValidation: vi.fn(async () => ({ rows: [], nextCursor: null })),
    recordDangling: vi.fn(async () => {}),
    ...overrides,
  };
}

const pdfFile: FileInput = {
  bytes: new Uint8Array(1024),
  name: "spec.pdf",
  mimeType: "application/pdf",
  sizeBytes: 1024,
};

const pngFile: FileInput = {
  bytes: new Uint8Array(2048),
  name: "photo.png",
  mimeType: "image/png",
  sizeBytes: 2048,
};

const thumbnailFile: FileInput = {
  bytes: new Uint8Array(256),
  name: "photo.thumb.jpg",
  mimeType: "image/jpeg",
  sizeBytes: 256,
};

describe("DefaultAttachmentPipeline.attach", () => {
  let storage: StorageProvider;
  let repo: AttachmentsRepo;
  let pipeline: DefaultAttachmentPipeline;

  beforeEach(() => {
    storage = makeStorage();
    repo = makeRepo();
    pipeline = new DefaultAttachmentPipeline({
      storage,
      repo,
      newClientId: () => CLIENT_ID,
    });
  });

  it("uploads bytes, inserts the row, returns the AttachmentRef", async () => {
    const ref = await pipeline.attach(NOTE, pdfFile);
    expect(ref.externalFileId).toBe("drive-original");
    expect(storage.upload).toHaveBeenCalledTimes(1);
    expect(repo.insert).toHaveBeenCalledTimes(1);
  });

  it("generates a thumbnail for image MIME types and uploads both", async () => {
    let uploadCount = 0;
    const storageWithCounter = makeStorage({
      upload: vi.fn(async (file: FileInput, hint?: string) => {
        uploadCount += 1;
        const externalId = hint === "thumbnail" ? "drive-thumb" : "drive-original";
        return makeRemote(externalId, file);
      }),
    });
    const thumbnails: ThumbnailGenerator = {
      generate: vi.fn(async () => thumbnailFile),
    };
    const p = new DefaultAttachmentPipeline({
      storage: storageWithCounter,
      repo,
      thumbnails,
      newClientId: () => CLIENT_ID,
    });
    const ref = await p.attach(NOTE, pngFile);
    expect(uploadCount).toBe(2);
    expect(thumbnails.generate).toHaveBeenCalledTimes(1);
    expect(ref.thumbnailId).toBe("drive-thumb");
  });

  it("does NOT generate a thumbnail for non-image MIME types", async () => {
    const thumbnails: ThumbnailGenerator = {
      generate: vi.fn(async () => thumbnailFile),
    };
    const p = new DefaultAttachmentPipeline({
      storage,
      repo,
      thumbnails,
      newClientId: () => CLIENT_ID,
    });
    await p.attach(NOTE, pdfFile);
    expect(thumbnails.generate).not.toHaveBeenCalled();
  });

  it("rolls back the uploaded bytes when the DB insert fails", async () => {
    const failingRepo = makeRepo({
      insert: vi.fn(async () => {
        throw new Error("RLS denied");
      }),
    });
    const p = new DefaultAttachmentPipeline({
      storage,
      repo: failingRepo,
      newClientId: () => CLIENT_ID,
    });
    await expect(p.attach(NOTE, pdfFile)).rejects.toThrow(/RLS denied/);
    expect(storage.delete).toHaveBeenCalledWith("drive-original");
  });

  it("rolls back both files when DB insert fails after the thumbnail upload", async () => {
    const storageWithThumb = makeStorage({
      upload: vi.fn(async (file: FileInput, hint?: string) => {
        const externalId = hint === "thumbnail" ? "drive-thumb" : "drive-original";
        return makeRemote(externalId, file);
      }),
      delete: vi.fn(async () => {}),
    });
    const thumbnails: ThumbnailGenerator = {
      generate: vi.fn(async () => thumbnailFile),
    };
    const failingRepo = makeRepo({
      insert: vi.fn(async () => {
        throw new Error("RLS denied");
      }),
    });
    const p = new DefaultAttachmentPipeline({
      storage: storageWithThumb,
      repo: failingRepo,
      thumbnails,
      newClientId: () => CLIENT_ID,
    });
    await expect(p.attach(NOTE, pngFile)).rejects.toThrow();
    expect(storageWithThumb.delete).toHaveBeenCalledWith("drive-original");
    expect(storageWithThumb.delete).toHaveBeenCalledWith("drive-thumb");
  });

  it("uploads original even if the thumbnail upload fails (soft-failure)", async () => {
    let uploads = 0;
    const flakyStorage = makeStorage({
      upload: vi.fn(async (file: FileInput, hint?: string) => {
        uploads += 1;
        if (hint === "thumbnail") {
          throw new Error("Drive 503 during thumbnail");
        }
        return makeRemote("drive-original", file);
      }),
    });
    const thumbnails: ThumbnailGenerator = {
      generate: vi.fn(async () => thumbnailFile),
    };
    const p = new DefaultAttachmentPipeline({
      storage: flakyStorage,
      repo,
      thumbnails,
      newClientId: () => CLIENT_ID,
    });
    const ref = await p.attach(NOTE, pngFile);
    expect(uploads).toBe(2);
    // No thumbnailId on the inserted row when the thumb upload failed.
    expect(ref.thumbnailId).toBeUndefined();
  });

  it("uploads original even if the thumbnail GENERATOR fails (soft-failure)", async () => {
    const flakyThumbnails: ThumbnailGenerator = {
      generate: vi.fn(async () => {
        throw new Error("Canvas exploded");
      }),
    };
    const p = new DefaultAttachmentPipeline({
      storage,
      repo,
      thumbnails: flakyThumbnails,
      newClientId: () => CLIENT_ID,
    });
    const ref = await p.attach(NOTE, pngFile);
    expect(ref.thumbnailId).toBeUndefined();
    expect(storage.upload).toHaveBeenCalledTimes(1);
  });

  it("publishes Failed when the original upload fails and does not call repo.insert", async () => {
    const failingStorage = makeStorage({
      upload: vi.fn(async () => {
        throw new Error("Drive 503");
      }),
    });
    const p = new DefaultAttachmentPipeline({
      storage: failingStorage,
      repo,
      newClientId: () => CLIENT_ID,
    });
    await expect(p.attach(NOTE, pdfFile)).rejects.toThrow(/Drive 503/);
    expect(repo.insert).not.toHaveBeenCalled();
    const snap = p["bus"].current(CLIENT_ID);
    expect(snap?.state.kind).toBe("Failed");
  });
});

describe("DefaultAttachmentPipeline.uploadProgress", () => {
  it("emits Pending -> Uploading -> Done across a successful upload", async () => {
    const storage = makeStorage();
    const repo = makeRepo();
    const pipeline = new DefaultAttachmentPipeline({
      storage,
      repo,
      newClientId: () => CLIENT_ID,
    });
    const seen: Progress[] = [];
    const iter = pipeline.uploadProgressByClientId(CLIENT_ID)[Symbol.asyncIterator]();
    const pump = (async () => {
      for (let i = 0; i < 3; i += 1) {
        const n = await iter.next();
        if (n.done) break;
        seen.push(n.value);
      }
    })();
    await pipeline.attach(NOTE, pdfFile);
    await pump;
    await iter.return?.();
    expect(seen.map((p) => p.state.kind)).toEqual(["Pending", "Uploading", "Done"]);
  });

  it("emits Failed on upload failure", async () => {
    const failingStorage = makeStorage({
      upload: vi.fn(async () => {
        throw new Error("boom");
      }),
    });
    const pipeline = new DefaultAttachmentPipeline({
      storage: failingStorage,
      repo: makeRepo(),
      newClientId: () => CLIENT_ID,
    });
    const seen: Progress[] = [];
    const iter = pipeline.uploadProgressByClientId(CLIENT_ID)[Symbol.asyncIterator]();
    const pump = (async () => {
      for (let i = 0; i < 3; i += 1) {
        const n = await iter.next();
        if (n.done) break;
        seen.push(n.value);
      }
    })();
    await expect(pipeline.attach(NOTE, pdfFile)).rejects.toThrow();
    await pump;
    await iter.return?.();
    expect(seen.at(-1)?.state.kind).toBe("Failed");
  });
});

describe("DefaultAttachmentPipeline.resolveUrl + remove", () => {
  it("resolveUrl proxies to StorageProvider.getDownloadUrl with the externalFileId", async () => {
    const storage = makeStorage();
    const pipeline = new DefaultAttachmentPipeline({ storage, repo: makeRepo() });
    const url = await pipeline.resolveUrl(sampleRef());
    expect(url).toBe("https://drive/drive-original");
    expect(storage.getDownloadUrl).toHaveBeenCalledWith("drive-original");
  });

  it("remove deletes the DB row first, then the remote bytes", async () => {
    const order: string[] = [];
    const storage = makeStorage({
      delete: vi.fn(async (id) => {
        order.push(`storage:${id}`);
      }),
    });
    const repo = makeRepo({
      delete: vi.fn(async (id) => {
        order.push(`db:${id}`);
      }),
    });
    const pipeline = new DefaultAttachmentPipeline({ storage, repo });
    await pipeline.remove(sampleRef({ thumbnailId: "drive-thumb" }));
    expect(order).toEqual([`db:${ATT}`, "storage:drive-original", "storage:drive-thumb"]);
  });

  it("remove swallows storage delete failures so the row stays gone", async () => {
    const storage = makeStorage({
      delete: vi.fn(async () => {
        throw new Error("Drive 503");
      }),
    });
    const repo = makeRepo();
    const pipeline = new DefaultAttachmentPipeline({ storage, repo });
    await expect(pipeline.remove(sampleRef())).resolves.toBeUndefined();
    expect(repo.delete).toHaveBeenCalled();
  });
});
